custom_components/tuya_local/devices/proscenic_a9_airpurifier.yaml

```diff
@@ -0,0 +1,149 @@
+name: Air Purifier
+products:
+  - id: giwqfeiezesbkkbe
+    name: Proscenic A9
+primary_entity:
+  entity: fan
+  icon: "mdi:air-purifier"
+  translation_key: fan_with_presets
+  dps:
+    - id: 1
+      name: switch
+      type: boolean
+    - id: 4
+      name: preset_mode
+      type: string
+      mapping:
+        - dps_val: "Sleep"
+          value: sleep
+        - dps_val: "1"
+          value: low
+        - dps_val: "2"
+          value: medium
+        - dps_val: "3"
+          value: high
+        - dps_val: "4"
+          value: turbo
+        - dps_val: "Auto"
+          value: auto
+    - id: 4
```

Comment:
Probably better to just use preset_mode, since all the options do not fit naturally into "speed"

Reply:
Well, the negative part of that is that, aside not being able to do stuff like increase fan speed commands, is that it makes impossible to directly start with a preset mode. 

In order to power on the device, instead of directly selecting a speed or a preset, you need to to turn on the fan and then select a preset mode from the home assistant interface from what i can see.

Else it will still report last preset mode but it actually goes in auto, which updates only by reopening the interface. It does a similar thing with the speed when just turning on the device with the power button in the same interface but at least it updates the interface saying auto and you can directly select a speed and it will turn at that specific mode directly.

If that's preferable i can surely change it but seems to reduce functionality. maybe mapping auto to 1% or something like that instead of 0%?

Reply:
fan.turn_on can take a preset_mode as an argument, and as of 2023.12.2, this integration will use that.

Reply:
I tried it by commenting out the speed part:

```
primary_entity:
  entity: fan
  icon: "mdi:air-purifier"
  translation_key: fan_with_presets
  dps:
    - id: 1
      name: switch
      type: boolean
    - id: 4
      name: preset_mode
      type: string
      mapping:
        - dps_val: "Sleep"
          value: sleep
        - dps_val: "1"
          value: low
        - dps_val: "2"
          value: medium
        - dps_val: "3"
          value: high
        - dps_val: "4"
          value: turbo
        - dps_val: "Auto"
          value: auto
#    - id: 4
#      name: speed
#      type: string
#      mapping:
#        - dps_val: "Sleep"
#          value: 20
#        - dps_val: "1"
#          value: 40
#        - dps_val: "2"
#          value: 60
#        - dps_val: "3"
#          value: 80
#        - dps_val: "4"
#          value: 100
#        - dps_val: "Auto"
#          value: 1
    - id: 102
      name: unknown_102
      type: string
```

but it has that behaviour:
if i choose a preset from HA nothing happens, if i turn on then the fan manually it keeps showing the preset i choose, but didn't actually set it and the device wakes up in auto mode. after several seconds it shows auto again. If I'm not doing anything wrong this pretty much doesn't make possible to directly turn on the fan on a specific preset.


https://github.com/make-all/tuya-local/assets/1127663/57ff8f19-0c38-4911-a98c-b3a5619055a3